bundles/CoreBundle/Controller/UserAwareController.php

```diff
@@ -0,0 +1,272 @@
+                    'user' => $this->getPimcoreUser()?->getName(),
+                    'permission' => $permission,
+                ]
+            );
+
+            throw $this->createAccessDeniedHttpException();
+        }
+    }
+
+    /**
+     * @param string $message
+     * @param \Throwable|null $previous
+     * @param int $code
+     * @param array $headers
+     *
+     * @return AccessDeniedHttpException
+     */
+    protected function createAccessDeniedHttpException(string $message = 'Access Denied.', \Throwable $previous = null, int $code = 0, array $headers = []): AccessDeniedHttpException
+    {
+        // $headers parameter not supported by Symfony 3.4
+        return new AccessDeniedHttpException($message, $previous, $code, $headers);
+    }
+
+    /**
+     * @param string[] $permissions
+     */
+    protected function checkPermissionsHasOneOf(array $permissions)
+    {
+        $allowed = false;
+        $permission = null;
+        foreach ($permissions as $permission) {
+            if ($this->getPimcoreUser()->isAllowed($permission)) {
+                $allowed = true;
+
+                break;
+            }
+        }
+
+        if (!$this->getPimcoreUser() || !$allowed) {
+            Logger::error(
+                'User {user} attempted to access {permission}, but has no permission to do so',
+                [
+                    'user' => $this->getPimcoreUser()->getName(),
+                    'permission' => $permission,
+                ]
+            );
+
+            throw new AccessDeniedHttpException('Attempt to access ' . $permission . ', but has no permission to do so.');
+        }
+    }
+
+    /**
+     * Check permission against all controller actions. Can optionally exclude a list of actions.
+     *
+     * @param ControllerEvent $event
+     * @param string $permission
+     * @param array $unrestrictedActions
+     */
+    protected function checkActionPermission(ControllerEvent $event, string $permission, array $unrestrictedActions = [])
+    {
+        $actionName = null;
+        $controller = $event->getController();
+
+        if (is_array($controller) && count($controller) === 2 && is_string($controller[1])) {
+            $actionName = $controller[1];
+        }
+
+        if (null === $actionName || !in_array($actionName, $unrestrictedActions)) {
+            $this->checkPermission($permission);
+        }
+    }
+
+    /**
+     * Returns a JsonResponse that uses the admin serializer
+     *
+     * @param mixed $data    The response data
+     * @param int $status    The status code to use for the Response
+     * @param array $headers Array of extra headers to add
+     * @param array $context Context to pass to serializer when using serializer component
+     * @param bool $usePimcoreSerializer
+     *
+     * @return JsonResponse
+     *
+     */
+    protected function jsonResponse($data, $status = 200, $headers = [], $context = [], bool $usePimcoreSerializer = true)
```

Comment:
I think we should move all the json related methods to a trait `Pimcore\Controller\Traits\JsonHelperTrait`. 
This we then can add to legacy `AdminController`  for BC and in Pimcore 11 we add it to those controllers needing this methods.

Reply:
we can move it to the trait however the catch is that every controller requiring it needs to inject the `pimcore.serailzer` service in service locator through `getSubscribedServices()`

Reply:
We could use the following in the trait: 

```php
    #[Required]
    public function setPimcoreSerializer($pimcoreSerializer)
    {
        $this->pimcoreSerializer = $pimcoreSerializer;
    }
```

Reply:
Ok. fixed with required injections https://github.com/pimcore/pimcore/pull/14918/commits/4ed583cfc70d89c0198a6c4780d0b2feba9368fe